### ISSUE TITLE
Proposed syntax on ‘require’

### ISSUE THREAD

**Issue by SpiffGreen** [midnqp/frameworkless#5]
First I'd like to say that the way requiring components is meant to be(in html) is a super nice idea and is highly supported, but the syntax may not look nice in the eye.
![image](https://user-images.githubusercontent.com/73044138/124395760-cd0dc380-dcfd-11eb-862f-efabd17daa83.jpeg)
So I'm proposing the another syntax for that
```html
<require src="/components/navbar" />
```
With support for passing of data through attributes.
```html
<require src="/components/board" name="John Doe" age=25>
```

**Comment by midnqp** [comment]
Well, I really wonder what useful data attribs for `require` could be?

**Comment by SpiffGreen** [comment]
Any attribute name can be given with it’s value. I’ve tested out a function for this, it will spit out an object {src: “/comps/board”, props: {name:”John”,..}}

**Comment by SpiffGreen** [comment]
The passing of props will be useful when using templating functionality, e.g
`<p>[name] is [age] years old</p>`

**Comment by SpiffGreen** [comment]
So what’s your take on this?
Your going for the syntax already in the specification or the one I’m proposing?

**Comment by midnqp** [comment]
I initially intended `require` (`frameworkless` in general) to be very simple like a language like `C`. So, I wanted `<require>` to just copy-paste whatever is inside `/public/comp/head.html`. And `head.html` may contain multiple `<script nodejs>`, which can run in the same `Execution Context` as the, for example, `/public/blog/view.html` - where `/public/blog/view.html` \<require\>'d the `head.html`.

Do you see my point? `Frameworkless` is meant to be the Node framework that has the lowest entry barrier. 

<details> 
<summary>Heavy breathe out</summary>

- In fact, every programming language should be simple, adequate tooling should make the language powerful. Like `Node`.
- SpiffGreen, I am so sorry that `midnqp\frameworkless` is not so active - I have many uncommitted changes. Because I got suddenly so dedicated in this [chatnet](https://github.com/midnqp/terminal.chat) project - using which anyone can communicate from a `Command Prompt` or terminal. It's not a new idea, but me designing+implementing the protocols feels like **wizardry**.

</details>

**Comment by SpiffGreen** [comment]
After careful considerations, I'll suggest that required scripts be executed before actually adding to the script that call them cos there could be multiple nested require statements which means a recursive style is needed.

**Comment by midnqp** [comment]
@SpiffGreen  Oh, that's a really well-thought consideration.

**Comment by SpiffGreen** [comment]
I talked about the syntax of require, so are we going to keep this syntax:
```
<require>/comp/nav.html</require>
```
Or you choose this:
```
<require src="/comp/nav" username="@spiff" >
```

**Comment by midnqp** [comment]
Let's keep `require` only for copy-pasting HTML.

<hr>

Personally, for templating I'd use functions. Put the individual files in `/public/templates/`. And inside a `<script nodejs>` of `/blog/view.html`,  I'd `require("/public/templates/databases")`, for templates/functions/a-block-of-repetitive-task I've written for interacting with database. Then just: 

`<template id=profile-display-min  
 sessionCookie="your session cookie string here"/>`. 

But you were right. Template will be almost like require. Difference is `require` deals with HTML, takes no data-attrib. `<template>` takes attrib, uses them, and returns the specific HTML to its place! 

Simple, right? Can you implement this?

**Comment by SpiffGreen** [comment]
Yes, I can implement it. And by the way it’s a good idea. I’ll be making a PR soon
